src/operator/batch_norm.cc

```diff
@@ -2,41 +2,492 @@
+                                                 const std::vector<TBlob> &out_data,
+                                                 const std::vector<TBlob> &aux_states) {
+  // Input
+  batchnorm::DeviceTensor3<DType> inputData(in_data[batchnorm::kData], 1);
+  const TBlob &weights         = in_data[batchnorm::kGamma];
+  const TBlob &bias            = in_data[batchnorm::kBeta];
+
+  // Aux (Moving)
+  const TBlob &runningMean     = aux_states[batchnorm::kMovingMean];
+  const TBlob &runningVariance = aux_states[batchnorm::kMovingVar];
+
+  // Output
+  batchnorm::DeviceTensor3<DType> outputData(out_data[batchnorm::kOut], 1);
+  const TBlob &meanVector      = out_data[batchnorm::kMean];
+  const TBlob &varianceVector  = out_data[batchnorm::kVar];
+
+  AccReal *mean = meanVector.dptr<AccReal>();
+  AccReal  *var = varianceVector.dptr<AccReal>();
+
+  const bool is_train_or_global_stats = ctx.is_train && !param_.use_global_stats;
+
+  if (is_train_or_global_stats) {
+    const TShape stride(2);
+
+    // compute mean per input
+    ComputeMean(inputData, meanVector.dptr<AccReal>());
+
+    // compute variance per input
+    ComputeVariance(inputData,
+                    meanVector.dptr<AccReal>(),
+                    static_cast<DType>(param_.eps),
+                    varianceVector.shape_,
+                    var);  // var is actually returned as invstd
+  } else {
+    const AccReal *rm = runningMean.dptr<AccReal>();
+    const AccReal *rv = runningVariance.dptr<AccReal>();
+
+    for (size_t i = 0, n = inputData.shape_[1]; i < n; ++i) {
+      mean[i] = rm[i];
+      var[i] = VARIANCE_TO_INVSTD(rv[i], param_.eps);
+    }
+  }
+
+  // compute output
+  AccReal        *w = weights.dptr<AccReal>();
+  const AccReal  *b = bias.dptr<AccReal>();
+
+    // note that var is still invstd
+    if (!param_.fix_gamma) {
+      if (IsWriting(req[batchnorm::kData])) {
+        ForEachFast(inputData, outputData,
+                    [w, b, mean, var](const size_t channel, const DType *in_data, DType *out_data) {
+                      *out_data = static_cast<DType>(
+                        ((*in_data - mean[channel]) * var[channel]) * w[channel] + b[channel]);
+                    });
+      }
+    } else {
+      if (IsWriting(req[batchnorm::kGamma])) {
+        for (size_t i =0, n = weights.Size(); i < n; ++i) {
+          w[i] = AccReal(1);
+        }
+      }
+      if (IsWriting(req[batchnorm::kData])) {
+        ForEachFast(inputData, outputData,
+                    [w, b, mean, var](const size_t channel, const DType *in_data, DType *out_data) {
+                      *out_data = static_cast<DType>(
+                        ((*in_data - mean[channel]) * var[channel]) + b[channel]);
+                    });
+      }
+    }
+
+  // Convert back to "real" variance in order to be consistent
+  // with the original operator
+  for (size_t i = 0, n = inputData.shape_[1]; i < n; ++i) {
+    var[i] = INVSTD_TO_VARIANCE(var[i], param_.eps);
```

Comment:
For performance, in forward pass, invstd is part of the per-item calculation.  Also, this is consistent with V1 for unit testing.
mshadow (V1) version converts liberally between the two.
To avoid possible cycling, I will adjust the forward pass a little to not convert the same value both ways